codegen/src/api/custom_values.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use std::collections::HashSet;
+
+use crate::{types::TypeGenerator, CratePath};
+use heck::ToSnakeCase as _;
+use subxt_metadata::Metadata;
+
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{format_ident, quote};
+
+/// Generate the custom values mod, if there are any custom values in the metadata. Else returns None.
+pub fn generate_custom_values(
+    metadata: &Metadata,
+    type_gen: &TypeGenerator,
+    types_mod_ident: &syn::Ident,
+    crate_path: &CratePath,
+) -> Option<TokenStream2> {
+    let mut fn_names_taken = HashSet::new();
+    let custom_value_fns: Vec<_> = metadata
+        .custom()
+        .iter()
+        .filter_map(|(key, custom)| {
+            generate_custom_value_fn(
+                key,
+                custom.type_id(),
+                type_gen,
+                crate_path,
+                &mut fn_names_taken,
+            )
+        })
+        .collect();
+
+    (!custom_value_fns.is_empty()).then(|| {
+        quote!(
+            pub fn custom() -> custom::CustomValuesApi {
+                custom::CustomValuesApi
+            }
+
+            pub mod custom{
+                use super::#types_mod_ident;
+
+                pub struct CustomValuesApi;
+
+                impl CustomValuesApi {
+                    #( #custom_value_fns )*
+                }
+            }
+
+
+        )
+    });
+
+    if custom_value_fns.is_empty() {
+        None
+    } else {
+        Some(quote!(
+            pub fn custom() -> custom::CustomValuesApi {
```

Comment:
With this implementation, I wonder if the metadata doesn't expose any custom values, then the users will not have the `fn custom()` exposed.

Wouldn't it be more consistent to have the `pub fn custom()` either way? Similar to how we expose `pub fn tx() -> TransactionApi`, then it would be an implementation detail that the `CustomValuesApi` doesn't expose anything